fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/FDBDirectoryLockFactory.java

```diff
@@ -0,0 +1,114 @@
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.IOException;
+
+/**
+ * Produce a lock over {@link FDBDirectory}.
+ */
+public final class FDBDirectoryLockFactory extends LockFactory {
+    private static final Logger LOGGER = LoggerFactory.getLogger(FDBDirectoryLockFactory.class);
+    final FDBDirectory directory;
+
+    public FDBDirectoryLockFactory(FDBDirectory directory) {
+        this.directory = directory;
+    }
+
+    @Override
+    public Lock obtainLock(final Directory dir, final String lockName) throws IOException {
+        // dir is ignored
+        return new FdbDirectoryLock(directory, lockName);
+    }
+
+    private static class FdbDirectoryLock extends Lock {
+
+        final FDBDirectory directory;
+        final byte[] key;
+        final byte[] value;
+        final String lockName;
+        final long timeStampMillis;
+        boolean closed;
+
+        public FdbDirectoryLock(final FDBDirectory directory, final String lockName) {
+            this.directory = directory;
+            this.lockName = lockName;
+            final Subspace subspace = directory.getSubspace();
+            this.key = subspace.pack("fileLock:" + lockName);
+            this.timeStampMillis = System.currentTimeMillis();
+            this.value = subspace.pack(this.timeStampMillis);
+            directory.fileLockSet(key, value);
+            tellMe("Set lock");
+        }
+
+        @Override
+        public void close() {
+            directory.fileLockClear(key, value);
+            closed = true;
+            tellMe("Release lock");
+        }
+
+        @Override
+        public void ensureValid() throws IOException {
+            if (closed) {
+                tellMe("validate: already close");
+                throw new AlreadyClosedException("Lock instance already released: " + this);
+            }
+            final Subspace subspace = directory.getSubspace();
+            byte[] currentValue = directory.fileLockGet(key);
+            if (currentValue == null) {
+                tellMe("validate: null lock");
+                throw new AlreadyClosedException("Lock file was deleted (lock=" + this + ")");
+            }
+            final long currentTimeStamp = subspace.unpack(currentValue).getLong(0);
+            if (currentTimeStamp != timeStampMillis) {
+                tellMe("validate: invalid lock", LogMessageKeys.TIME_STARTED_MILLIS, currentValue);
+                throw new AlreadyClosedException("Lock file changed by an external force at " + currentTimeStamp + ", (lock=" + this + ")");
+            }
+        }
+
+        @Override
+        public String toString() {
+            return "FDBDirectory: name=" + lockName + " timeMillis=" + timeStampMillis;
+        }
+
+        @Nonnull
+        private void tellMe(@Nonnull String staticMsg, @Nullable final Object... keysAndValues) {
+            if (false && LOGGER.isDebugEnabled()) {
```

Comment:
Is this temporary? Has a somewhat peculiar name and doesn't ever do anything.

Reply:
✅